### PR TITLE
HDDS-4984. Support async profiler 2.0

### PR DESCRIPTION
**Pull request by sky76093016**
## What changes were proposed in this pull request?

Due to the update of Async profiler 2.0, the output format is changed from svg to html.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4984

## How was this patch tested?

Local test.

### CODE REVIEW

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @sky76093016 for working on this.  Can you please keep SVG support, too?  For example `ozone-runner` docker image has Async Profiler 1.5.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java [hunk 1]
```diff
@@ -365,8 +378,8 @@ protected void doGetDownload(String fileName, final HttpServletRequest req,
           "This page will auto-refresh every 2 second until output file is "
               + "ready..");
     } else {
-      if (safeFileName.endsWith(".svg")) {
-        resp.setContentType("image/svg+xml");
+      if (safeFileName.endsWith(".html")) {
+        resp.setContentType("image/html+xml");
```
Comment (cku328) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java:382]:
```suggestion
        resp.setContentType("text/html");
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java [hunk 2]
```diff
@@ -500,7 +515,8 @@ public static Event fromInternalName(final String name) {
     COLLAPSED,
     SVG,
     TREE,
-    JFR
+    JFR,
+    HTML
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java:519]: Async Profiler 2.0 does not seem to create useful output with `-o html`:

```
$ /opt/profiler/profiler.sh -d 10 -o html -f /tmp/prof-output/async-prof-pid-7-cpu-1.html 7
Profiling for 10 seconds
Done
$ cat /tmp/prof-output/async-prof-pid-7-cpu-1.html
Stopped profiling after 10 seconds. No dump options specified
```

We need `-o flamegraph` instead:

```
$ /opt/profiler/profiler.sh -d 10 -o flamegraph -f /tmp/prof-output/async-prof-pid-7-cpu-2.html 7
Profiling for 10 seconds
Done
$ head -3 /tmp/prof-output/async-prof-pid-7-cpu-2.html
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg version="1.1" width="1200" height="518" onload="init(evt)" viewBox="0 0 1200 518" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @sky76093016 for updating the patch.  I think output format needs some tweak for 2.0.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java (hunk 2 above)
Comment (sky76093016) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java]: Thanks for tips. :")

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java (hunk 1 above)
Comment (sky76093016) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ProfileServlet.java]: Done.

**Comment by jojochuang** [comment]
@adoroszlai  is it? i think it threw an error at me when i specified svg output...
regardless, the html report is easier to use.

**Comment by jojochuang** [comment]
So... looks like even with the change, you would still need to point the browser to /prof?output=flamegraph in order to use async profiler 2.0, otherwise the servlet will attempt to generate svg output and that will fail.

**Comment by jojochuang** [comment]
how about adding another system property  to specify the async profiler version, and defaults to flamegraph if it's 2.0?

**Comment by adoroszlai** [comment]
> > 2.0 works with -o svg, although profiler's help does not mention it as a valid output format.
> is it? i think it threw an error at me when i specified svg output...

I think I was fooled by some docker-specific behavior (tried to replace `/opt/profiler` at runtime...).

Now with proper 2.0 installation I see the problem with `-o svg`.

It might be improved by checking the subprocess for error result instead of waiting forever for the output file.

> how about adding another system property to specify the async profiler version, and defaults to flamegraph if it's 2.0?

We could also run the profiler with `-v` first to get the version number.

**Review by jojochuang — APPROVED** [review]
+1 as discussed offline, let's merge this first, and we can work on the follow-up later.

**Review by cku328 — APPROVED** [review]
+1 for this PR. It is currently at least initially compatible with version 2.0.

**Comment by elek** [comment]
> +1 as discussed offline, let's merge this first, and we can work on the follow-up later.

Did we open the follow-up issues? Looks like ideal newbie issues....

**Comment by sky76093016** [comment]
Thanks @elek . I have issued jira and will finish it in a few days. The website of jira is: https://issues.apache.org/jira/browse/HDDS-5009

**Pull request merged**